### PR TITLE
KickPlanner now uses game state as a factor for determining a kick

### PR DESCRIPTION
**Pull request by CptTony57**

### CODE REVIEW

File: module/behaviour/planning/KickPlanner/src/KickPlanner.cpp [hunk 1]
```diff
@@ -92,12 +99,13 @@ namespace behaviour {
                 }
             });
 
-            on<Trigger<Ball>, With<Field>, With<FieldDescription>, With<KickPlan>, With<Sensors>>().then(
-                [this](const Ball& ball,
-                       const Field& field,
-                       const FieldDescription& fd,
-                       const KickPlan& kickPlan,
-                       const Sensors& sensors) {
+            on<Trigger<Ball>, With<Field>, With<FieldDescription>, With<KickPlan>, With<GameState>, With<Sensors>>()
```
Comment (TrentHouliston) [inline module/behaviour/planning/KickPlanner/src/KickPlanner.cpp]: Make the game state optional for this, we still want to be able to kick balls when testing without a game controller.

File: module/behaviour/planning/KickPlanner/src/KickPlanner.cpp [hunk 2]
```diff
@@ -23,6 +23,8 @@
 
 #include "message/behaviour/KickPlan.h"
 #include "message/behaviour/ServoCommand.h"
+#include "message/input/GameEvents.h"
```
Comment (TrentHouliston) [inline module/behaviour/planning/KickPlanner/src/KickPlanner.cpp]: Don't think you need this header

**Review by TrentHouliston — DISMISSED** [review]
Make sure this also disables it from kicking if we are in the free kick ready state. We don't want to get yellow cards.

**Review by JosephusPaye — APPROVED** [review]
Code LGTM

File: module/behaviour/planning/KickPlanner/src/KickPlanner.cpp [hunk 3]
```diff
@@ -50,12 +52,18 @@ namespace behaviour {
         using message::behaviour::KickPlan;
         using KickType = message::behaviour::KickPlan::KickType;
         using message::behaviour::WantsToKick;
+        using message::input::GameState;
         using message::input::Sensors;
         using message::localisation::Ball;
         using message::localisation::Field;
+        using Phase         = message::input::GameState::Data::Phase;
+        using PenaltyReason = message::input::GameState::Data::PenaltyReason;
```
Comment (KipHamiltons) [inline module/behaviour/planning/KickPlanner/src/KickPlanner.cpp:60]: wouldn't `using message::input::GameState::Data::Phase` have the same effect as the explicit assignment of the alias? 
same for the line after
```suggestion
        using message::input::GameState::Data::Phase;
        using message::input::GameState::Data::PenaltyReason;
```

File: module/behaviour/planning/KickPlanner/src/KickPlanner.cpp [hunk 4]
```diff
@@ -92,12 +100,20 @@ namespace behaviour {
                 }
             });
 
-            on<Trigger<Ball>, With<Field>, With<FieldDescription>, With<KickPlan>, With<Sensors>>().then(
-                [this](const Ball& ball,
-                       const Field& field,
-                       const FieldDescription& fd,
-                       const KickPlan& kickPlan,
-                       const Sensors& sensors) {
+            on<Trigger<ButtonMiddleDown>, Single>().then([this] { forcePlaying = true; });
```
Comment (KipHamiltons) [inline module/behaviour/planning/KickPlanner/src/KickPlanner.cpp:103]: How do you turn it off? could we do something like `forcePlaying = !forcePlaying`?

File: module/behaviour/planning/KickPlanner/src/KickPlanner.hpp [hunk 5]
```diff
@@ -36,6 +36,7 @@ namespace behaviour {
 
         private:
             bool kickValid(const arma::vec3& ballPos);
```
Comment (KipHamiltons) [inline module/behaviour/planning/KickPlanner/src/KickPlanner.hpp:38]: what is this even doing? is it saying that if the arma::vec is null then it's false?? 
could be out of the scope of this PR, but this line makes no sense to me

File: module/behaviour/planning/KickPlanner/src/KickPlanner.cpp (hunk 4 above)
Comment (JosephusPaye) [reply module/behaviour/planning/KickPlanner/src/KickPlanner.cpp:103]: Middle button always plays. Turning it off is red button.

**Review by KipHamiltons — APPROVED** [review]
LGTM

**Pull request merged**